momento/simple_cache_client.go

```
@@ -7,140 +7,144 @@ import (
 	"github.com/momentohq/client-sdk-go/internal/services"
 )
 
-const DeafultDataCtxTimeout = 5
+const defaultRequestTimeout = uint32(5)
 
-type ScsClient struct {
+type ScsClient interface {
+	CreateCache(request *CreateCacheRequest) error
+	DeleteCache(request *DeleteCacheRequest) error
+	ListCaches(request *ListCachesRequest) (*ListCachesResponse, error)
+
+	Set(request *CacheSetRequest) (*SetCacheResponse, error)
+	Get(request *CacheGetRequest) (*GetCacheResponse, error)
+
+	Close() error
+}
+
+type DefaultScsClient struct {
 	authToken         string
 	defaultTtlSeconds uint32
 	controlClient     *services.ScsControlClient
 	dataClient        *services.ScsDataClient
 }
 
-func SimpleCacheClient(request *SimpleCacheClientRequest) (*ScsClient, error) {
+func NewSimpleCacheClient(request *SimpleCacheClientRequest) (ScsClient, error) {
 	endpoints, err := resolver.Resolve(&models.ResolveRequest{
 		AuthToken: request.AuthToken,
 	})
 	if err != nil {
-		return nil, NewMomentoError(err.Code(), err.Message(), err.OriginalErr())
+		return nil, convertMomentoSvcErrorToCustomerError(err)
 	}
 	controlClient, err := services.NewScsControlClient(&models.ControlClientRequest{
 		AuthToken: request.AuthToken,
 		Endpoint:  endpoints.ControlEndpoint,
 	})
 	if err != nil {
-		err = momentoerrors.ConvertSvcErr(err)
-		return nil, NewMomentoError(err.Code(), err.Message(), err.OriginalErr())
+		return nil, convertMomentoSvcErrorToCustomerError(momentoerrors.ConvertSvcErr(err))
 	}
-	err = validateRequestTimeout(request.RequestTimeoutSeconds)
-	if err != nil {
-		return nil, NewMomentoError(err.Code(), err.Message(), err.OriginalErr())
+
+	requestTimeoutToUse := defaultRequestTimeout
+	if request.RequestTimeoutSeconds > 1 {
+		requestTimeoutToUse = request.RequestTimeoutSeconds
 	}
```

I believe Python and other SDKs throw an exception if the provided timeout is 0 or less.
For out case, we're setting the default timeout if the provided timeout is 0 or 1 without warning the user?

---

Hmm so first off good catch. What I was intending here was `> 0` not 1.  I felt the optional *unit32 added more pain then was worth. So switched to just uint32 So now it defaults to 0 under the hood if not set and its impossible to tell if user passed or if it just wasnt set. Before logic was set to only fail if was less then 1  and because we use uint32 for type now they cant pass negative number or something bad.

I think the scenario that concerns me is user explicitly passes 0 expecting idk what exactly and then we just silently use default request timeout under the hood....

curious maybe @gautamomento thoughts here also. I can try adding back to throw error if they pass 0 but we will need to go back to *uint32 then vs uint32 pretty sure.

---

Yes, so the reason why I used a pointer to pass (even though it felt weird since that was the only field in the struct to pass as a pointer) to distinguish actual 0 timeout passed by the user and default 0 set by the struct.

>  user explicitly passes 0 expecting idk what exactly

I just followed other SDK patters to check if user-provided timeout or default timeout is 0 but I felt the same as your comment.

---

idk I think my preference would be to just keep this not as pointer and wait and see if any real customers have issue with this approach? Could be convinced otherwise for sure though. Lets let @gautamomento make final call since he has been owning e2e sdk experience. Im fine either way but slight preference to no pointer I think after playing w/ sdk last night.

---

Would it be any different if we used some `time.Duration` class assuming that it is a first class citizen in Golang? That way we have clear indication of whether it is provided or not. 

I have chosen Duration class in Java as it is a first class citizen  - customers can do whatever they prefer millis, nanos, seconds, hours etc. I would prefer at some point to make our ttls also comply to this but that is a topic for some other time.

---

It seems that `time.Duration` results in `int64` so if the user doesn't pass anything in `timeout` field, it'll be 0 and still not way for us to distinguish the user provided 0 from default value 0?
https://pkg.go.dev/time#Duration

---

I am not sure what the right way out here. Not entering anything and using defaults make sense to me. 

Entering zero (even when, you shouldn't be) and then using defaults just seems wrong to me.  I am in favor of whatever mechanism we need to use to not make this happen.

---

@eaddingtonwhite With Gautam's feedback, do we want to go back to the pointer `timeout` so that we can check if the user passed 0 or nil?

---

per our zoom call have changed up how this works per that convo. Going to resolve this thread and we can open new comments on new approach.